contracts/amm/pool/Joseph.sol

```diff
@@ -15,31 +15,37 @@ abstract contract Joseph is JosephInternal, IJoseph {
     using SafeCast for int256;
 
     function calculateExchangeRate() external view override returns (uint256) {
-        return _calculateExchangeRate(block.timestamp);
+        IMiltonInternal milton = _getMilton();
+        return
+            _calculateExchangeRate(
+                block.timestamp,
+                milton,
+                _getIpToken(),
+                milton.getAccruedBalance().liquidityPool
+            );
     }
 
     function provideLiquidity(uint256 assetAmount) external override whenNotPaused {
-        _provideLiquidity(assetAmount, _getDecimals(), block.timestamp);
+        _provideLiquidity(assetAmount, block.timestamp);
     }
 
     function redeem(uint256 ipTokenAmount) external override whenNotPaused {
         _redeem(ipTokenAmount, block.timestamp);
     }
 
-    function checkVaultReservesRatio() external view override returns (uint256) {
-        return _checkVaultReservesRatio();
-    }
-
-    function _calculateExchangeRate(uint256 calculateTimestamp) internal view returns (uint256) {
-        IMiltonInternal milton = _getMilton();
-
+    function _calculateExchangeRate(
+        uint256 calculateTimestamp,
+        IMiltonInternal milton,
```

Comment:
Wouldn't it be better to pass soap instead of calculateTimestamp and milton?

Reply:
You right. I will try fix it.